### ISSUE TITLE
"undefined reference to `_flsbuf'" when compiling with ucrt64 gcc and using flto

### ISSUE THREAD

**Issue by Morilli** [msys2/MINGW-packages#8968]
I noticed that when compiling certain programs using ucrt64 gcc instead of mingw64 or mingw32 gcc *with -flto*, they will error on linking with the message `undefined reference to '_flsbuf'`.
Here's a simple test program that should reproduce this behavior when compiled with a ucrt64 gcc and using a simple compile command just as `gcc -Wall -Wextra -flto test.c -o test.exe`:
```c
// test.c
#include <stdio.h>
#include <stdlib.h>

int main()
{
    FILE* temp_file = fopen("__test_file__.txt", "wb");
    if (!temp_file) exit(EXIT_FAILURE);

    _lock_file(temp_file);
    for (int i = 0; i < 10000; i++) {
        _putc_nolock(i % 256, temp_file);
    }
    fclose(temp_file);
}
```

This seems to *only* happen when combining
1. ucrt64 gcc
2. compiling with `-flto` given
3. using stdio `_*_nolock` functions such as `_putc_nolock`

The full exact output from running the above command for me is:
```
C:/msys64/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:/msys64/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/lib/../lib/libmsvcrt.a(dyocs00098.o):(.text+0x0): multiple definition of `_putc_nolock'; C:/msys64/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/lib/../lib/libmingwex.a(lib64_libmingwex_a-_putc_nolock.o):C:/_/M/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/stdio/_putc_nolock.c:7: first defined here
C:/msys64/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:/msys64/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/lib/../lib/libmingwex.a(lib64_libmingwex_a-_putc_nolock.o): in function `_putc_nolock':
C:/_/M/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/stdio/_putc_nolock.c:7: undefined reference to `_flsbuf'
collect2.exe: error: ld returned 1 exit status
```

**Comment by Biswa96** [comment]
Also why does it try to link with libmsvcrt.a? /cc @lhmouse

**Comment by jeremyd2019** [comment]
what libmsvcrt actually is is controlled by the configure option.  on ucrt it is actually ucrt, but named msvcrt so tools continue to work.

**Comment by lhmouse** [comment]
@mstorsjo UCRT has all `_{get,put}{,w}c_nolock` functions, so probably they should be compiled in libmingwex.

**Comment by mstorsjo** [comment]
> @mstorsjo UCRT has all `_{get,put}{,w}c_nolock` functions, so probably they shouldn't be compiled in libmingwex.

Thanks, noted. I'll look into making a patch for this (but I'm a bit slow at the moment so it'll probably be a couple days).

**Comment by mstorsjo** [comment]
I don't have a suitable setup right now to reproduce + test a fix, but I sent a patch to mingw-w64-public which moves those relevant functions from libmingwex.a to libmsvcr*.a, which I presume would fix this issue. (If you try out the patch, note that it requires rerunning "autoreconf" in the mingw-w64-crt directory after applying it.)

**Comment by mstorsjo** [comment]
I pushed the tentative fix for this now, we'll see after the mingw-w64-crt package for ucrt64 gets updated with the latest git version if the fix worked out for this case.

**Comment by mstorsjo** [comment]
Ok, I tested building a GCC cross compiler targeting UCRT, and I was able to reproduce the bug with an older version of mingw-w64, and the fix I pushed does indeed fix the issue.

**Comment by Morilli** [comment]
Fixed now with 12c0067383b6ddd361082f6a1858757701269d12, thank you!
